src/sources/amqp.rs

```diff
@@ -0,0 +1,565 @@
+
+                                            // Error is logged by `codecs::Decoder`, no further handling
+                                            // is needed here.
+                                            if !error.can_continue() {
+                                                break;
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                            .boxed();
+
+                            match finalizer {
+                                Some(ref finalizer) => {
+                                    let (batch, receiver) = BatchNotifier::new_with_receiver();
+                                    let mut stream = stream.map(|event| event.with_batch_notifier(&batch));
+
+                                    match out.send_event_stream(&mut stream).await {
+                                        Err(error) => {
+                                            emit!(StreamClosedError { error, count: 1 });
+                                        }
+                                        Ok(_) => {
+                                            // TODO Is this needed?
+                                            // drop(stream);
+                                            finalizer.add(msg.into(), receiver);
+                                        }
+                                    }
+                                }
+                                None => {
+                                    match out.send_event_stream(&mut stream).await {
+                                        Err(error) => {
+                                            emit!(StreamClosedError { error, count: 1 });
+                                        }
+                                        Ok(_) => {
+                                            let ack_options = lapin::options::BasicAckOptions::default();
+                                            if let Err(error) = msg.acker.ack(ack_options).await {
+                                                error!(message = "Unable to ack.", error = ?error, internal_log_rate_secs = 10);
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    break
+                }
+            }
+        };
+    }
+
+    Ok(())
+}
+
+async fn handle_ack(status: BatchStatus, entry: FinalizerEntry) {
+    match status {
+        BatchStatus::Delivered => {
+            let ack_options = lapin::options::BasicAckOptions::default();
+            if let Err(error) = entry.acker.ack(ack_options).await {
+                error!(message = "Unable to ack.", error = ?error, internal_log_rate_secs = 10);
```

Comment:
For each of the `error!` in here, should these be an internal event per the instrumentation spec ?